Mathlib/Analysis/CStarAlgebra/CStarMatrix.lean

```diff
@@ -0,0 +1,489 @@
+  set Dn := Fintype.card n
+  set Dm := Fintype.card m
+  simp [Matrix.norm_def, Pi.norm_def, Pi.nnnorm_def]
+  by_cases hm_triv : Nonempty m
+  · by_cases hn_triv : Nonempty n
+    · obtain ⟨i, _, hi⟩ := exists_mem_eq_sup (univ : Finset m) (univ_nonempty_iff.mpr hm_triv)
+        fun b => Finset.univ.sup fun b_1 => ‖toMatrixAux M b b_1‖₊
+      obtain ⟨j, _, hj⟩ := exists_mem_eq_sup (univ : Finset n) (univ_nonempty_iff.mpr hn_triv)
+        fun b_1 => ‖toMatrixAux M i b_1‖₊
+      rw [hi, hj]
+      change ‖M‖₊ ≤ ↑Dn * ↑Dm * ‖M i j‖₊
+      refine nnnorm_le_of_forall_inner_le fun v w => ?_
+      simp only [WithCStarModule.pi_inner, CStarMatrix.toCLM_apply_eq_sum, mul_sum]
+      have hmax : ∀ k l, ‖M k l‖₊ ≤ ‖M i j‖₊ := fun k l => by
+        change (univ.sup fun b => univ.sup fun b_1 => ‖M b b_1‖₊)
+          = univ.sup fun b_1 => ‖M i b_1‖₊ at hi
+        change (univ.sup fun b_1 => ‖M i b_1‖₊) = ‖M i j‖₊ at hj
+        calc ‖M k l‖₊ ≤ univ.sup fun l' => ‖M k l'‖₊ :=
+                  Finset.le_sup (f := fun l' => ‖M k l'‖₊) (mem_univ l)
+          _ ≤ univ.sup fun k' => univ.sup fun l' => ‖M k' l'‖₊ :=
+                  Finset.le_sup (f := fun k' => univ.sup fun l' => ‖M k' l'‖₊) (mem_univ k)
+          _ = ‖M i j‖₊ := by rw [← hj, ← hi]
+      calc _ ≤ ∑ k, ‖∑ l, star (w k) * M k l * v l‖₊ := by
+                  simp_rw [mul_assoc]
+                  refine (nnnorm_sum_le (E := A) _ _).trans ?_
+                  gcongr
+                  simp [WithCStarModule.inner_def, mul_sum]
+        _ ≤ ∑ k, ∑ l, ‖star (w k) * M k l * v l‖₊ := by gcongr; exact nnnorm_sum_le _ _
+        _ ≤ ∑ k, ∑ l, ‖star (w k) * M k l‖₊ * ‖v l‖₊ := by gcongr; exact nnnorm_mul_le _ _
+        _ ≤ ∑ k, ∑ l, ‖w k‖₊ * ‖M k l‖₊ * ‖v l‖₊ := by
+                  gcongr with k _ l _
+                  refine (nnnorm_mul_le _ _).trans_eq ?_
+                  simp_rw [nnnorm_star (w k)]
+        _ ≤ ∑ k, ∑ l, ‖w k‖₊ * ‖M i j‖₊ * ‖v l‖₊ := by gcongr with k _ l _; exact hmax k l
+        _ = ∑ k, ∑ l, ‖M i j‖₊ * (‖w k‖₊ * ‖v l‖₊) := by
+                  congr 1; ext k; norm_cast
+                  congr 1; ext l; norm_cast
+                  ring
+        _ = ‖M i j‖₊ * (∑ k, ∑ l, ‖w k‖₊ * ‖v l‖₊) := by simp [← mul_sum]
+        _ = (∑ k, ∑ l, ‖w k‖₊ * ‖v l‖₊) * ‖M i j‖₊ := by rw [mul_comm]
+        _ ≤ (∑ (_ : m), ∑ (_ : n), ‖w‖₊ * ‖v‖₊) * ‖M i j‖₊ := by
+                  gcongr <;> exact WithCStarModule.norm_apply_le_norm _ _
+        _ = (Dm * (Dn * (‖w‖₊ * ‖v‖₊))) * ‖M i j‖₊ := by congr; simp [sum_const]
+        _ = Dn * Dm * ‖M i j‖₊ * ‖v‖₊ * ‖w‖₊ := by ring
+    · simp only [not_nonempty_iff] at hn_triv
+      simp only [Finset.sup_eq_bot_of_isEmpty, Finset.sup_bot]
+      simp only [bot_eq_zero', NNReal.coe_zero, mul_zero, norm_le_zero_iff]
+      ext _ j
+      exact False.elim <| IsEmpty.false j
+  · simp only [not_nonempty_iff] at hm_triv
+    simp [Finset.sup_eq_bot_of_isEmpty, bot_eq_zero]
+    ext i j
+    exact False.elim <| IsEmpty.false i
```

Comment:
This needs, or else the proof breaks, see for yourself.
```lean
attribute [-instance] CStarMatrix.instNorm -- to avoid Lean seeing this instance on `Matrix m n A`?
```
This actually seems like a big problem. Lean sees the `CStarMatrix` norm on `Matrix` because it's an `abbrev`, even when the bit in question is `toMatrixAux A`, which is declared to have type `Matrix m n A`.

It's possible that this won't come up enough in our own work to matter, and probably in other places in the library Lean would have no business inferring `CStarMatrix` in place of `Matrix`, *but* I could imagine people would be pretty upset if they were using their `Matric n n ℂ` and all of a sudden our `CStarMatrix` norm came waltzing in.

It's possible we can't use `abbrev`. What do you think?

```suggestion
open CStarMatrix WithCStarModule in
private lemma antilipschitzWith_toMatrixAux :
    AntilipschitzWith (Fintype.card m * Fintype.card n)
      (toMatrixAux : CStarMatrixAux m n A → Matrix m n A) := by
  refine AddMonoidHomClass.antilipschitz_of_bound _ fun M => ?_
  calc
    ‖M‖ ≤ ∑ i, ∑ j, ‖M i j‖ := by
      rw [norm_def]
      refine (toCLM A M).opNorm_le_bound (by positivity) fun v => ?_
      simp only [toCLM_apply_eq_sum, equiv_symm_pi_apply, Finset.sum_mul]
      apply pi_norm_le_sum_norm _ |>.trans
      gcongr with i _
      apply norm_sum_le _ _ |>.trans
      gcongr with j _
      apply norm_mul_le _ _ |>.trans
      gcongr
      exact norm_apply_le_norm v j
    _ ≤ ∑ _ : m, ∑ _ : n, ‖toMatrixAux M‖ := by
      gcongr with i _ j _
      exact toMatrixAux M |>.norm_entry_le_entrywise_sup_norm
    _ = _ := by simp [mul_assoc]
```

Reply:
I agree this is horrible. I didn't think Lean would try to go "backwards" like this to find an instance (meaning going to `CStarMatrix` to find something about `Matrix`). I'll switch the whole thing from `abbrev` to `def` when I get a bit of time.

Reply:
I didn't think it would do that either. This sucks. ☹️